File: 백태균/week10/43105_정수_삼각형.java
class Solution {
    public int solution(int[][] triangle) {
        int[][] dp = new int[triangle.length][triangle.length];

        dp[0][0] = triangle[0][0];  // 초기값

        for (int i=1; i<triangle.length; i++) {
            for (int j=0; j<triangle[i].length; j++) {
                if (j == 0) {  // 왼쪽 대각선은 왼쪽 대각선 그대로 더한 값을 dp에 저장
                    dp[i][j] = triangle[i][j] + dp[i-1][j];
                } else if (j == triangle[i].length-1) {  // 오른쪽 대각선도 오른쪽 대각선 그대로 더한 값을 dp에 저장
                    dp[i][j] = triangle[i][j] + dp[i-1][j-1];
                } else {  // 그 외에는 해당 지점에서 왼쪽 아래, 오른쪽 아래 각각 더한 값의 최대값을 dp에 저장
                    dp[i][j] = triangle[i][j] + Math.max(dp[i-1][j-1], dp[i-1][j]);
                }
            }
        }

//         for (int i=0; i<triangle.length; i++) {
//             for (int j=0; j<triangle[i].length; j++) {
//                 System.out.print(dp[i][j] + " ");
//             }
//             System.out.println();
//         }

        int result = 0;
        for (int i=0; i<dp[triangle.length-1].length; i++) {
            result = Math.max(result, dp[triangle.length-1][i]);
        }

        return result;
    }
}

File: 김진효/week10/43105_정수_삼각형.py
"""
[성능 요약] 메모리: 15.8 MB 시간: 19.75 ms 

정확성: 64.3
효율성: 35.7
합계: 100.0 / 100.0

"""

# def solution(triangle):
#     prev = triangle[0]

#     for i in range(1,len(triangle)):
#         cur = triangle[i]
#         temp = [0] * len(cur)
#         for j in range(len(cur)):
#             temp[j] = cur[j]
#             left = j - 1
#             right = j
#             # 범위 고려
#             if left < 0 :
#                 temp[j] += prev[right] 
#             elif right == len(cur) - 1:
#                 temp[j] += prev[left]
#             else:
#                 temp[j] += max(prev[left], prev[right])
        
#         prev = temp
            

#     return max(prev)


#  그냥 기존 triangle 활용하기
 # [성능 요약] 메모리: 15.3 MB 시간: 25.33 ms 
def solution(triangle):
    for i in range(1,len(triangle)):
        cur = triangle[i]
        for j in range(len(cur)):
            left = j - 1
            right = j
            # 범위 고려
            if left < 0 :
                triangle[i][j] += triangle[i-1][right] 
            elif right == len(cur) - 1:
                triangle[i][j] += triangle[i-1][left]
            else:
                triangle[i][j] += max(triangle[i-1][left], triangle[i-1][right])
        
    return max(triangle[-1])

print(solution([[7], [3, 8], [8, 1, 0], [2, 7, 4, 4], [4, 5, 2, 6, 5]]))